SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestRepNotify/SpatialTestRepNotify.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestRepNotify.h"
+#include "SpatialGDK/Public/EngineClasses/SpatialNetDriver.h"
+
+#include "Net/UnrealNetwork.h"
+
+/**
+ * This test tests RepNotifies and shadow data implementation.
+ *
+ * The test includes 1 Server and 2 Client workers.
+ * The flow is as follows:
+ * - Setup:
+ *	- The test itself is used as the test Actor.
+ *  - It contains a number of replicated properties and OnRep functions that either set boolean flags confirming they have been called, or save the old value of the argument passed to the OnRep function.
+ * - Test:
+ *	- The Server changes the values of the replicated properties multiple times.
+ *  - The clients check whether the RepNotifies have or have not been called correctly.
+ * - Cleanup:
+ *	- None.
+ */
+
+ASpatialTestRepNotify::ASpatialTestRepNotify()
+	: Super()
+{
+	Author = "Miron";
+	Description = TEXT("Test RepNotify replication and shadow data");
+}
+
+void ASpatialTestRepNotify::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// The Server sets some initial values to the replicated variables.
+	AddStep(TEXT("SpatialTestRepNotifyServerSetReplicatedVariables"), FWorkerDefinition::Server(1), nullptr, [this]()
+		{
+			TestInt1 = 1;
+			TestInt2 = 2;
+			TestInt3 = 3;
+			TestInt4 = 4;
+			TestArray.Empty();
+			TestArray.Add(1);
+			TestArray.Add(2);
+
+			FinishStep();
+		});
+
+	// All clients check they received the correct values for the replicated variables.
+	AddStep(TEXT("SpatialTestRepNotifyAllClientsCheckReplicatedVariables"), FWorkerDefinition::AllClients, nullptr, nullptr, [this](float DeltaTime)
+		{
+			if (bOnRepTestInt1Called && bOnRepTestInt2Called
+				&& TestInt1 == 1 && TestInt2 == 2 && TestInt3 == 3 && TestInt4 == 4
+				&& TestArray.Num() == 2 && TestArray[0] == 1 && TestArray[1] == 2)
+			{
+				FinishStep();
+			}
+		}, 5.0f);
+
+	// All clients reset the values modified by the RepNotifies called due to the Server modifying the replicated variables.
+	AddStep(TEXT("SpatialTestRepNotifiyAllClientsLocallyChangeVariables"), FWorkerDefinition::AllClients, nullptr, [this]()
+		{
+			bOnRepTestInt1Called = false;
+			bOnRepTestInt2Called = false;
+			OldTestInt3 = -3;
+			OldTestInt4 = -4;
+			OldTestArray.Empty();
+
+			FinishStep();
+		});
+
+	// All clients modify 3 of the replicated variables.
+	AddStep(TEXT("SpatialTestRepNotifyAllClientsModifyReplicatedVariables"), FWorkerDefinition::AllClients, [this]() -> bool
+		{
+			// Extra check to ensure that the Old values are correct before changing the replicated variables.
+			if (OldTestInt3 == -3 && OldTestInt4 == -4)
```

Comment:
@mironec  this was part of your original test, I can't see why would it be necessary though, can't think of a situation where those would not have the correct value, apart from a possibly impossible race condition?

Reply:
Yeah, I think this is not possible. Can be removed.

Reply:
Will do, thanaank you!